packages/contracts-sdk/src/implementations/farcaster/FarcasterKeyGatewayContract.ts

```diff
@@ -0,0 +1,248 @@
+import {
+  SIGNED_KEY_REQUEST_VALIDATOR_EIP_712_TYPES,
+  KEY_GATEWAY_EIP_712_TYPES,
+} from "@farcaster/hub-nodejs";
+import {
+  EVMAccountAddress,
+  EVMContractAddress,
+  BlockchainCommonErrors,
+  UnixTimestamp,
+  FarcasterKeyGatewayContractError,
+  FarcasterId,
+  SignedKeyRequestSignature,
+  EncodedSignedKeyRequestMetadata,
+  SignedAddSignature,
+} from "@snickerdoodlelabs/objects";
+import { ParamType, TypedDataField, ethers } from "ethers";
+import { injectable } from "inversify";
+import { ResultAsync, okAsync } from "neverthrow";
+
+import { BaseContract } from "@contracts-sdk/implementations/BaseContract.js";
+import { IEthersContractError } from "@contracts-sdk/implementations/BlockchainErrorMapper.js";
+import {
+  ContractOverrides,
+  IFarcasterKeyGatewayContract,
+  WrappedTransactionResponse,
+} from "@contracts-sdk/interfaces/index.js";
+import {
+  ContractsAbis,
+  SignedKeyRequest,
+} from "@contracts-sdk/interfaces/objects/index.js";
+
+@injectable()
+export class FarcasterKeyGatewayContract
+  extends BaseContract<FarcasterKeyGatewayContractError>
+  implements IFarcasterKeyGatewayContract
+{
+  constructor(
+    protected providerOrSigner: ethers.Provider | ethers.Signer,
+    protected contractAddress: EVMContractAddress,
+  ) {
+    super(
+      providerOrSigner,
+      contractAddress,
+      ContractsAbis.FarcasterKeyGatewayAbi.abi,
+    );
+  }
+
+  public nonces(
+    address: EVMAccountAddress,
+  ): ResultAsync<
+    bigint,
+    FarcasterKeyGatewayContractError | BlockchainCommonErrors
+  > {
+    return ResultAsync.fromPromise(
+      this.contract.nonces(address) as Promise<bigint>,
+      (e) => {
+        return this.generateError(e, "Unable to call nonces()");
+      },
+    );
+  }
+
+  public add(
+    key: string,
+    encodedMetadata: string, // obtained from getEncodedSignedKeyRequestMetadata()
```

Comment:
If this is a string, we should definitely have a primitive type, FarcasterEncodedSignedKeyRequestMetadata